workflow/controller/scope.go

```diff
@@ -57,26 +60,71 @@ func (s *wfScope) resolveVar(v string) (interface{}, error) {
 	return nil, errors.Errorf(errors.CodeBadRequest, "Unable to resolve: {{%s}}", v)
 }
 
-func (s *wfScope) resolveParameter(v string) (string, error) {
-	val, err := s.resolveVar(v)
+func (s *wfScope) resolveParameter(p *wfv1.ValueFrom) (string, error) {
+	if p == nil {
+		return "", nil
+	}
+	if p.Parameter == "" && p.Expression == "" {
+		return "", nil
+	}
+	var val interface{}
+	var err error
+	if p.Expression != "" {
+		val, err = expr.Eval(p.Expression, s.getAllParamArtifact())
+	} else {
+		val, err = s.resolveVar(p.Parameter)
+	}
+
 	if err != nil {
-		return "", err
+		return "", fmt.Errorf("unable to resolve expression: %s", err)
 	}
-	valStr, ok := val.(string)
-	if !ok {
-		return "", errors.Errorf(errors.CodeBadRequest, "Variable {{%s}} is not a string", v)
+	if val == nil {
+		return "", nil
 	}
-	return valStr, nil
+	return fmt.Sprintf("%v", val), nil
 }
 
-func (s *wfScope) resolveArtifact(v string, subPath string) (*wfv1.Artifact, error) {
-	val, err := s.resolveVar(v)
+func (s *wfScope) getAllParamArtifact() map[string]interface{} {
+
+	paramArtMap := make(map[string]interface{})
+	for key, val := range s.scope {
+		if _, ok := val.(*wfv1.AnyString); ok {
+			paramArtMap[key] = val.(*wfv1.AnyString).Value()
+		} else {
+			paramArtMap[key] = val
+		}
+	}
+	for _, param := range s.tmpl.Inputs.Parameters {
+		key := fmt.Sprintf("inputs.parameters.%s", strings.TrimSpace(param.Name))
+		paramArtMap[key] = s.tmpl.Inputs.GetParameterByName(param.Name).Value.Value()
+	}
+	for _, param := range s.tmpl.Inputs.Artifacts {
+		key := fmt.Sprintf("inputs.artifacts.%s", param.Name)
+		paramArtMap[key] = s.tmpl.Inputs.GetArtifactByName(param.Name)
+	}
+	return paramArtMap
+}
+
+func (s *wfScope) resolveArtifact(art *wfv1.Artifact, subPath string) (*wfv1.Artifact, error) {
+	if art == nil {
+		return nil, nil
+	}
+	if art.From == "" && art.FromExpression == "" {
+		return nil, nil
+	}
+	var err error
+	var val interface{}
+	if art.FromExpression != "" {
+		val, err = expr.Eval(art.FromExpression, s.getAllParamArtifact())
+	} else {
+		val, err = s.resolveVar(art.From)
+	}
 	if err != nil {
-		return nil, err
+		return nil, fmt.Errorf("unable to resolve artifact: %s", err)
 	}
 	valArt, ok := val.(wfv1.Artifact)
```

Comment:
I don't see how `val` could ever be an `Artifact`?

Reply:
Scope map will have artifacts.